File: src/aetherius/core/runtime/flow.py
"""Flow actions (``if``/``repeat``/``for_each``/``optional``): interpreted before any driver.

Split out of the step executor (steps.py), which re-enters itself through the ``FlowHost``
protocol for the nested lists — so every Act inherits the flow semantics without wiring
anything, and the executor file stays focused on the pipeline itself.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Callable, Protocol, Sequence

from pydantic import ValidationError

from ..actions.base import FLOW_NESTED_FIELDS, Capability
from ..blueprint.models import StepModel
from ..errors import ActionError
from .result import RunStatus

if TYPE_CHECKING:
    from .context import RunContext

_TRUTHY: frozenset[str] = frozenset({"true", "1", "yes"})

# Template context names a for_each loop variable would shadow (see RunContext.template_ctx).
_RESERVED_NAMES: frozenset[str] = frozenset({"inputs", "secrets", "vars", "env", "steps"})

_DEFAULT_LOOP_VAR = "item"


def is_truthy(value: Any) -> bool:
    """Single truthiness rule shared by ``assert`` conditions and ``when`` guards."""
    return str(value).strip().lower() in _TRUTHY


@dataclass(frozen=True)
class FlowOutcome:
    """What a flow step reports back: its outputs, and the status the executor must record.

    Only ``optional`` ever reports anything but SUCCESS. Carrying the status here rather than
    inferring it in the executor keeps a single rule — the step that *interpreted* the block says
    how it went.
    """

    outputs: dict[str, Any] = field(default_factory=dict)
    status: RunStatus = RunStatus.SUCCESS


class FlowHost(Protocol):
    """The slice of the executor a flow action needs: recursion plus the run scope."""

    ctx: "RunContext"

    def run(self, steps: Sequence[StepModel], path: str = "", act: str | None = None) -> None: ...

    def run_tolerant(self, steps: Sequence[StepModel], path: str, act: str) -> bool: ...


def run_flow(
    host: FlowHost,
    step: StepModel,
    renderer: Callable[[Any], Any],
    path: str,
    act: str,
) -> FlowOutcome:
    """Interpret one flow step; nested lists re-enter *host* with the step's effective *act*."""
    if step.action == "if":
        return FlowOutcome(_flow_if(host, step, renderer, path, act))
    if step.action == "repeat":
        return FlowOutcome(_flow_repeat(host, step, renderer, path, act))
    # Explicit, because for_each is the fallback below: without this branch an `optional` block
    # would be interpreted as a malformed loop.
    if step.action == Capability.OPTIONAL.value:
        return _flow_optional(host, step, path, act)
    return FlowOutcome(_flow_for_each(host, step, renderer, path, act))


def _flow_optional(host: FlowHost, step: StepModel, path: str, act: str) -> FlowOutcome:
    """Run a block whose failure is an acceptable outcome (Jalon 3-J).

    The block publishes no data of its own: what happened is read from ``Result.step_results``,
    never from the template context — the same split the milestone installs everywhere.
    """
    nested = _nested_steps(step, "steps", path)
    yielded = host.run_tolerant(nested, path, act)
    return FlowOutcome({}, RunStatus.PARTIAL if yielded else RunStatus.SUCCESS)


def _flow_if(
    host: FlowHost, step: StepModel, renderer: Callable[[Any], Any], path: str, act: str
) -> dict[str, Any]:
    p = step.extra_fields
    if "condition" not in p:
        raise ActionError(f"if: missing required parameter 'condition' (step {path!r}).")
    branch = "then" if is_truthy(renderer(p["condition"])) else "else"
    if branch == "else" and "else" not in p:
        return {"branch": None}
    host.run(_nested_steps(step, branch, path), path, act)
    return {"branch": branch}


def _flow_repeat(
    host: FlowHost, step: StepModel, renderer: Callable[[Any], Any], path: str, act: str
) -> dict[str, Any]:
    p = step.extra_fields
    if "times" not in p:
        raise ActionError(f"repeat: missing required parameter 'times' (step {path!r}).")
    times = _coerce_times(renderer(p["times"]), path)
    nested = _nested_steps(step, "steps", path)
    for i in range(times):
        host.run(nested, f"{path}[{i}]", act)
    return {"iterations": times}


def _flow_for_each(
    host: FlowHost, step: StepModel, renderer: Callable[[Any], Any], path: str, act: str
) -> dict[str, Any]:
    p = step.extra_fields
    if "items" not in p:
        raise ActionError(f"for_each: missing required parameter 'items' (step {path!r}).")
    items = renderer(p["items"])
    if not isinstance(items, (list, tuple)):
        raise ActionError(
            f"for_each: 'items' must render to a list, got {type(items).__name__} (step {path!r})."
        )
    var = p.get("as", _DEFAULT_LOOP_VAR)
    if not isinstance(var, str) or not var.isidentifier():
        raise ActionError(
            f"for_each: 'as' must be a valid identifier, got {var!r} (step {path!r})."
        )
    if var in _RESERVED_NAMES:
        raise ActionError(
            f"for_each: loop variable {var!r} would shadow a reserved "
            f"template name (step {path!r})."
        )
    nested = _nested_steps(step, "steps", path)

    # Save and restore whatever the variable shadowed so nested loops compose.
    missing = object()
    previous = host.ctx.scope.get(var, missing)
    try:
        for i, item in enumerate(items):
            host.ctx.scope[var] = item
            host.run(nested, f"{path}[{i}]", act)
    finally:
        if previous is missing:
            host.ctx.scope.pop(var, None)
        else:
            host.ctx.scope[var] = previous
    return {"iterations": len(items)}


def seed_block_outputs(ctx: "RunContext", step: StepModel) -> None:
    """Publish ``{}`` for every identified step of an ``optional`` block that produced nothing.

    Both engines reject the undefined at the point of *use*, so ``steps.coord.ville | default(null)``
    raises when ``steps.coord`` is missing altogether — the filter never sees the value, the
    attribute access already failed. Seeding an empty dict is what makes the documented writing rule
    true. ``setdefault`` is the whole subtlety: a step that published before giving way keeps what
    it published.

    It recurses through nested blocks so the rule holds at any depth inside the block, and never
    touches anything outside it. The consequence is deliberate and documented: ``steps.coord is
    defined`` now holds even when the block gave way — the template context carries *data*, while
    ``Result.step_results`` carries what *happened*.
    """
    _seed(ctx, step.id, step.action, step.extra_fields)


def _seed(ctx: "RunContext", step_id: Any, action: str, fields: dict[str, Any]) -> None:
    if isinstance(step_id, str) and step_id:
        ctx.step_outputs.setdefault(step_id, {})
    for name in FLOW_NESTED_FIELDS.get(action, ()):
        nested = fields.get(name)
        if not isinstance(nested, list):
            continue
        for item in nested:
            if isinstance(item, dict):
                _seed(ctx, item.get("id"), str(item.get("action", "")), item)


def _nested_steps(step: StepModel, key: str, path: str) -> list[StepModel]:
    raw = step.extra_fields.get(key)
    if not isinstance(raw, list):
        raise ActionError(f"{step.action}: '{key}' must be a list of steps (step {path!r}).")
    try:
        return [StepModel.model_validate(item) for item in raw]
    except ValidationError as exc:
        raise ActionError(f"{step.action}: invalid step in '{key}' (step {path!r}): {exc}") from exc


def _coerce_times(value: Any, path: str) -> int:
    times: int | None = None
    if isinstance(value, int) and not isinstance(value, bool):
        times = value
    elif isinstance(value, str):
        try:
            times = int(value.strip())
        except ValueError:
            times = None
    if times is None:
        raise ActionError(f"repeat: 'times' must be an integer, got {value!r} (step {path!r}).")
    if times < 0:
        raise ActionError(f"repeat: 'times' must be >= 0, got {times} (step {path!r}).")
    return times

File: sdks/engine/src/runtime/flow.ts
/**
 * Flow actions (`if` / `repeat` / `for_each` / `optional`), mirror of
 * `src/aetherius/core/runtime/flow.py`.
 *
 * They are interpreted before any driver sees them, so every Act inherits the semantics without
 * wiring anything. The executor re-enters itself through `FlowHost` for the nested lists.
 *
 * **The loops are sequential `await`s, never `Promise.all`.** Turning them into parallel work
 * because "we are async now" would make runs non-reproducible and break every Blueprint whose
 * iterations share a session or read the previous one's outputs. The order of steps is observable
 * — it is in the event stream — so it is part of the contract, not an implementation detail.
 */

import { nestedStepFields } from "../blueprint/contract.js";
import type { StepModel } from "../blueprint/types.js";
import type { Renderer, RunContext } from "../driver.js";
import { ActionError } from "../errors.js";
import { isTruthy, pythonRepr } from "../expr/index.js";
import type { RunStatus } from "../result.js";

/** Template names a `for_each` loop variable would shadow (see `templateContext`). */
const RESERVED_NAMES = new Set(["inputs", "secrets", "vars", "env", "steps"]);

const DEFAULT_LOOP_VAR = "item";

/**
 * ASCII only, where Python's `str.isidentifier()` accepts any Unicode letter.
 *
 * Deliberately narrower: Unicode property escapes (`\p{L}`) are not guaranteed by the mobile JS
 * engine, and a regexp literal that fails to compile takes the whole module down at load time —
 * a crash on a device instead of a rejected loop variable. Refusing more than Python does is the
 * safe direction here, and it is the one the phase already assumes (a strict subset). Documented
 * in docs/embedded.md.
 */
const IDENTIFIER = /^[A-Za-z_][A-Za-z0-9_]*$/;

/**
 * What a flow step reports back: its outputs, and the status the executor must record.
 *
 * Only `optional` ever reports anything but `success`. Carrying the status here rather than
 * inferring it in the executor keeps a single rule — the step that *interpreted* the block says
 * how it went.
 */
export interface FlowOutcome {
  readonly outputs: Record<string, unknown>;
  readonly status: RunStatus;
}

/** The slice of the executor a flow action needs: recursion, plus the run scope. */
export interface FlowHost {
  readonly ctx: RunContext;
  run(steps: readonly StepModel[], path: string, act: string): Promise<void>;
  runTolerant(steps: readonly StepModel[], path: string, act: string): Promise<boolean>;
}

/** Interpret one flow step; nested lists re-enter *host* with the step's effective *act*. */
export async function runFlow(
  host: FlowHost,
  step: StepModel,
  render: Renderer,
  path: string,
  act: string,
): Promise<FlowOutcome> {
  if (step.action === "if") return done(await flowIf(host, step, render, path, act));
  if (step.action === "repeat") return done(await flowRepeat(host, step, render, path, act));
  // Explicit, because `for_each` is the fallback below: without this branch an `optional` block
  // would be interpreted as a malformed loop, and fail on a missing `items`.
  if (step.action === "optional") return flowOptional(host, step, path, act);
  return done(await flowForEach(host, step, render, path, act));
}

function done(outputs: Record<string, unknown>): FlowOutcome {
  return { outputs, status: "success" };
}

/**
 * Run a block whose failure is an acceptable outcome (milestone 3-J).
 *
 * The block publishes no data of its own: what happened is read from `Result.step_results`, never
 * from the template context — the same split the milestone installs everywhere.
 */
async function flowOptional(
  host: FlowHost,
  step: StepModel,
  path: string,
  act: string,
): Promise<FlowOutcome> {
  const nested = nestedSteps(step, "steps", path);
  const gaveWay = await host.runTolerant(nested, path, act);
  return { outputs: {}, status: gaveWay ? "partial" : "success" };
}

/**
 * Publish `{}` for every identified step of an `optional` block that produced nothing.
 *
 * Both engines reject the undefined at the point of *use*, so `steps.coord.city | default(null)`
 * throws when `steps.coord` is missing altogether — the filter never sees the value, the member
 * access already failed. Seeding an empty object is what makes the documented writing rule true.
 * `??=` is the whole subtlety: a step that published before giving way keeps what it published.
 *
 * It recurses through nested blocks so the rule holds at any depth inside the block, and never
 * touches anything outside it. The consequence is deliberate and documented: `steps.coord` is
 * defined even when the block gave way — the template context carries *data*, while
 * `Result.step_results` carries what *happened*.
 */
export function seedBlockOutputs(ctx: RunContext, step: StepModel): void {
  seed(ctx, step.id, step.action, step as Record<string, unknown>);
}

function seed(ctx: RunContext, stepId: unknown, action: string, fields: Record<string, unknown>): void {
  if (typeof stepId === "string" && stepId !== "") ctx.stepOutputs[stepId] ??= {};
  for (const name of nestedStepFields(action)) {
    const nested = fields[name];
    if (!Array.isArray(nested)) continue;
    for (const item of nested) {
      if (item !== null && typeof item === "object") {
        const child = item as Record<string, unknown>;
        seed(ctx, child["id"], typeof child["action"] === "string" ? child["action"] : "", child);
      }
    }
  }
}

async function flowIf(
  host: FlowHost,
  step: StepModel,
  render: Renderer,
  path: string,
  act: string,
): Promise<Record<string, unknown>> {
  if (!has(step, "condition")) {
    throw new ActionError(`if: missing required parameter 'condition' (step '${path}').`);
  }
  const branch = isTruthy(render(step["condition"])) ? "then" : "else";
  if (branch === "else" && !has(step, "else")) return { branch: null };
  await host.run(nestedSteps(step, branch, path), path, act);
  return { branch };
}

async function flowRepeat(
  host: FlowHost,
  step: StepModel,
  render: Renderer,
  path: string,
  act: string,
): Promise<Record<string, unknown>> {
  if (!has(step, "times")) {
    throw new ActionError(`repeat: missing required parameter 'times' (step '${path}').`);
  }
  const times = coerceTimes(render(step["times"]), path);
  const nested = nestedSteps(step, "steps", path);
  for (let i = 0; i < times; i += 1) {
    await host.run(nested, `${path}[${i}]`, act);
  }
  return { iterations: times };
}

async function flowForEach(
  host: FlowHost,
  step: StepModel,
  render: Renderer,
  path: string,
  act: string,
): Promise<Record<string, unknown>> {
  if (!has(step, "items")) {
    throw new ActionError(`for_each: missing required parameter 'items' (step '${path}').`);
  }
  const items = render(step["items"]);
  if (!Array.isArray(items)) {
    throw new ActionError(
      `for_each: 'items' must render to a list, got ${typeName(items)} (step '${path}').`,
    );
  }
  const variable = has(step, "as") ? step["as"] : DEFAULT_LOOP_VAR;
  if (typeof variable !== "string" || !IDENTIFIER.test(variable)) {
    throw new ActionError(
      `for_each: 'as' must be a valid identifier, got ${pythonRepr(variable)} (step '${path}').`,
    );
  }
  if (RESERVED_NAMES.has(variable)) {
    throw new ActionError(
      `for_each: loop variable '${variable}' would shadow a reserved template name ` +
        `(step '${path}').`,
    );
  }
  const nested = nestedSteps(step, "steps", path);

  // Save and restore whatever the variable shadowed, so nested loops compose.
  const { scope } = host.ctx;
  const had = Object.prototype.hasOwnProperty.call(scope, variable);
  const previous = scope[variable];
  try {
    for (let i = 0; i < items.length; i += 1) {
      scope[variable] = items[i];
      await host.run(nested, `${path}[${i}]`, act);
    }
  } finally {
    if (had) scope[variable] = previous;
    else delete scope[variable];
  }
  return { iterations: items.length };
}

function nestedSteps(step: StepModel, key: string, path: string): StepModel[] {
  const raw = step[key];
  if (!Array.isArray(raw)) {
    throw new ActionError(`${step.action}: '${key}' must be a list of steps (step '${path}').`);
  }
  return raw.map((item, index) => {
    if (item === null || typeof item !== "object" || typeof (item as StepModel).action !== "string") {
      throw new ActionError(
        `${step.action}: invalid step in '${key}' at index ${index} (step '${path}'): ` +
          "each entry must be an object with an 'action'.",
      );
    }
    return item as StepModel;
  });
}

function coerceTimes(value: unknown, path: string): number {
  let times: number | null = null;
  if (typeof value === "number" && Number.isInteger(value)) {
    times = value;
  } else if (typeof value === "string" && /^[+-]?\d+$/.test(value.trim())) {
    times = Number(value.trim());
  }
  if (times === null) {
    throw new ActionError(
      `repeat: 'times' must be an integer, got ${pythonRepr(value)} (step '${path}').`,
    );
  }
  if (times < 0) {
    throw new ActionError(`repeat: 'times' must be >= 0, got ${times} (step '${path}').`);
  }
  return times;
}

/** Python-flavoured type name, so the two engines' messages stay recognisable side by side. */
function typeName(value: unknown): string {
  if (value === null || value === undefined) return "NoneType";
  if (Array.isArray(value)) return "list";
  if (typeof value === "string") return "str";
  if (typeof value === "boolean") return "bool";
  if (typeof value === "number") return Number.isInteger(value) ? "int" : "float";
  return "dict";
}

function has(step: StepModel, key: string): boolean {
  return Object.prototype.hasOwnProperty.call(step, key);
}
